File: packages/mpcore/lib/mpkit/web_view.dart
part of 'mpkit.dart';

class MPWebViewController extends MPPlatformViewController {
  void reload() {
    invokeMethod('reload', requireResult: true);
  }

  void loadUrl(String url) {
    invokeMethod('loadUrl', params: {'url': url});
  }
}

class MPWebView extends MPPlatformView {
  final String url;
  final Function(List<dynamic>)? onMiniProgramMessage;

  @override
  final MPWebViewController? controller;

  MPWebView({required this.url, this.controller, this.onMiniProgramMessage})
      : super(
            viewType: 'mp_web_view',
            viewAttributes: {'url': url},
            controller: controller,
            onMethodCall: (method, arguments) {
              if (method == 'mini_program_message' &&
                  arguments is Map &&
                  arguments['data'] is List) {
                if (onMiniProgramMessage != null) {
                  onMiniProgramMessage.call(arguments['data'] as List);
                }
              }
            });
}

File: packages/mp_dom_runtime/src/components/mpkit/web_view.ts
import { MPEnv } from "../..";
import { PlatformType } from "../../env";
import { setDOMAttribute, setDOMStyle } from "../dom_utils";
import { MPPlatformView } from "./platform_view";

export class MPWebView extends MPPlatformView {
  constructor(readonly document: Document, readonly initialAttributes?: any) {
    super(document, initialAttributes);
    if (__MP_MINI_PROGRAM__) {
      this.htmlElement.addEventListener("message", (e: any) => {
        if (e?.detail?.data instanceof Array) {
          this.invokeMethod("mini_program_message", { data: e.detail.data });
        }
      });
    }
  }

  elementType() {
    if (__MP_MINI_PROGRAM__) {
      return "wx-web-view";
    }
    return "iframe";
  }

  onMethodCall(method: string, params: any) {
    if (method === "reload") {
      setDOMAttribute(this.htmlElement, "src", "");
      setDOMAttribute(this.htmlElement, "src", this.attributes.url);
    } else if (method === "loadUrl" && params?.url) {
      setDOMAttribute(this.htmlElement, "src", "");
      setDOMAttribute(this.htmlElement, "src", params.url);
    }
  }

  setAttributes(attributes: any) {
    super.setAttributes(attributes);
    setDOMAttribute(this.htmlElement, "src", attributes.url);
    setDOMStyle(this.htmlElement, { border: "none" });
  }

  setChildren() {}
}
